haskell-auto-import.el

```diff
@@ -0,0 +1,286 @@
+          ;; Optionally match the qualified keyword as group 1
+          "\\(qualified\\s-+\\)?"
+          ;; Match exactly the module name
+          (regexp-quote module-name)
+          ;; Optionally match an `as' qualifier as group 2
+          "\\(?:\\s-+as\\s-+\\(\\sw+\\)\\)?"
+          ;; Whitespace may or may not precede parentheses at the end
+          "\\s-*"
+          ;; Optionally match the contents of some following parentheses as
+          ;; group 3
+          "\\(?:(\\(\\(?:[a-zA-Z0-9', \t\f\n\r]*\\|(.*?)\\)*\\))\\)?"
+          )))
+    (if (string-match regexp contents start)
+        (let ((is-qualified (match-string 1 contents))
+              (qualified-name (match-string 2 contents))
+              (imported-items-str (match-string 3 contents))
+              (start-point (match-beginning 0))
+              (end-point (match-end 0)))
+          (list start-point end-point
+                (if is-qualified t nil)
+                qualified-name
+                (if imported-items-str
+                    (haskell-auto-import-parse-import-parens imported-items-str)
+                  nil)
+                )))))
+
+(defun haskell-auto-import-module-import-info (module-name &optional start)
+  "Returns information about every import of the given module after `start'
+(which is set to 0 if nil). Each member of the returned list has the same format
+as that returned from `haskell-auto-import-module-single-import-info'."
+  (let ((next-info (haskell-auto-import-module-single-import-info module-name start)))
+    (if next-info
+        (let ((next-start (car (cdr next-info))))
+          (cons next-info (haskell-auto-import-module-import-info module-name next-start)))
+      ()
+      )))
+
+(defun haskell-auto-import-normalize-imported-items (imported-items)
+  "Normalize imported items, e.g. `one(a), two(a, b), two(a, c), one' --> `one,
+two(a, b, c)'"
+  (let ((grouped (-group-by 'car imported-items)))
+    (mapcar
+     '(lambda (item-group)
+        (let* ((group-main (car item-group))
+               (all-subs (mapcar 'cdr (cdr item-group)))
+               (flat-subs (reduce 'append all-subs))
+               (distinct-subs (-distinct flat-subs)))
+          (if (member nil all-subs)
+              (list group-main)
+            (cons group-main distinct-subs)
+            )))
+     grouped)))
+
+(defun haskell-auto-import-print-imported-items (imported-items)
+  (mapconcat
+   '(lambda (item)
+      (let ((imported-item (car item))
+            (sub-imports (cdr item)))
+        (concat
+         imported-item
+         (if sub-imports
+             (concat
+              "("
+              (mapconcat 'identity sub-imports ", ")
+              ")")
+             ""))
+        ))
+   imported-items ", "))
+
+(defun haskell-auto-import-print-module-import (module-name info)
+  (let ((is-qualified (car info))
+        (qualified-name (car (cdr info)))
+        (imported-items (car (cdr (cdr info)))))
+    (concat
```

Comment:
`format` might be easier to read than `concat` here if you let-bind some of the `if` expressions.